larq_zoo/train.py

```diff
@@ -1,65 +1,132 @@
-    import larq as lq
-    from larq_zoo import utils
-    import tensorflow as tf
-
-    initial_epoch = utils.get_current_epoch(output_dir)
-    model_path = path.join(output_dir, "model")
-    callbacks = [utils.ModelCheckpoint(filepath=model_path, save_weights_only=True)]
-    if hasattr(hparams, "learning_rate_schedule"):
-        callbacks.append(
-            tf.keras.callbacks.LearningRateScheduler(hparams.learning_rate_schedule)
-        )
-    if tensorboard:
-        callbacks.append(
-            tf.keras.callbacks.TensorBoard(log_dir=output_dir, write_graph=False)
-        )
 
-    with tf.device("/cpu:0"):
-        train_data = dataset.train_data(hparams.batch_size)
-        validation_data = dataset.validation_data(hparams.batch_size)
+class TrainLarqZooModel(Experiment):
+    # Save model checkpoints.
+    use_model_checkpointing: bool = Field(True)
 
-    with utils.get_distribution_scope(hparams.batch_size):
-        model = build_model(hparams, **dataset.preprocessing.kwargs)
-        model.compile(
-            optimizer=hparams.optimizer,
-            loss="categorical_crossentropy",
-            metrics=["categorical_accuracy", "top_k_categorical_accuracy"],
+    # Log metrics to Tensorboard.
+    use_tensorboard: bool = Field(True)
+
+    # Use a per-batch progress bar (as opposed to per-epoch).
+    use_progress_bar: bool = Field(False)
+
+    # Where to store output.
+    @Field
+    def output_dir(self) -> Union[str, os.PathLike]:
+        return (
+            Path.home()
+            / "zookeeper-logs"
+            / self.dataset.__class__.__name__
+            / self.__class__.__name__
+            / datetime.now().strftime("%Y%m%d_%H%M")
         )
-        lq.models.summary(model)
-
-        if initial_epoch > 0:
-            model.load_weights(model_path)
-            click.echo(f"Loaded model from epoch {initial_epoch}")
-
-    model.fit(
-        train_data,
-        epochs=hparams.epochs,
-        steps_per_epoch=dataset.train_examples // hparams.batch_size,
-        validation_data=validation_data,
-        validation_steps=dataset.validation_examples // hparams.batch_size,
-        verbose=2 if tensorboard else 1,
-        initial_epoch=initial_epoch,
-        callbacks=callbacks,
-    )
 
-    model_name = build_model.__name__
-    model.save_weights(path.join(output_dir, f"{model_name}_weights.h5"))
+    @property
+    def model_path(self):
+        return self.output_dir / "model"
 
-    # Save weights without top
-    notop_model = build_model(
-        hparams, **dataset.preprocessing.kwargs, include_top=False
+    metrics: List[Union[Callable[[tf.Tensor, tf.Tensor], float], str]] = Field(
+        lambda: ["sparse_categorical_accuracy", "sparse_top_k_categorical_accuracy"]
     )
-    notop_model.set_weights(model.get_weights()[: len(notop_model.get_weights())])
-    notop_model.save_weights(path.join(output_dir, f"{model_name}_weights_notop.h5"))
+
```

Comment:
```suggestion
learning_rate_schedule: Optional[Callable] = None
```

Since `BiRealNet` currently does not have an LR schedule and the callbacks check `if self.learning_rate_schedule`, training BRN will throw an error. Also, should we change those checks to `if self.learning_rate_schedule is not None` as with object detection?

Reply:
Nice, thanks :+1: